### ISSUE TITLE
questions about details

### ISSUE THREAD

**Issue by ChenzrMax** [Turoad/CLRNet#4]
Thanks for your great work! Some questions confused me:
1. Does the method predict fixed number of lanes with anchor-like lane prior(x,y,theta)? And what's the number of the lane pirors in your experiment? Or it's just an anchor-based method like Line-CNN and LaneATT?

2.In section "Refinement structure", "Pt is the parameter of lane prior (start point coordinate x, y and angle θ)", does it mean you only refine (x,y,theta) and only use (x,y,theta) to extract lane features, and use ROIalign to extract 1x1xC feature at Np sampled positions?

3. In section "Refinement structure", "The refinement Rt takes the Pt as input to get the ROI lane features and then performs two FC layers to get the refined parameter Pt", does it mean you only use the ROI lane features to predict x,y,theta? and which feature do you use to predict confidence, length, and N offsets? and what's the shape of the output(multi head output or just use FC layers to get M x (2+1+3+N))?

4. in section "ROIGather structure", "For ROI features of L1, L2, we concatenate the ROI features of previous layers to enhance feature representations. Convolutions are performed on the extracted ROI features to gather nearby features for each lane pixel.", along which dimension do you concatenate the features and what's the shape of the convolution kernel.

5. in section "Inference", "Our method can also be nms-free if we use the one-to-one assignment, i.e., set the top-k = 1.", does it mean assign the top-1 predicted lane when training or just select one predicted lane by classification score when inference?

hope the code coming soon

**Comment by ChenzrMax** [comment]
Excuse me, I have another question about Tensorrt.
Do you export the pytorch model to onnx model and then use tensorrt to speed up? If so, which opset of onnx do you use since I find ROIAlign only supported in opset 10 and 16.

Hope for your answer.

**Comment by MySuperSoul** [comment]
Thanks for your interest with our paper. As for your questions, I will explain to you in below.

Q1: As for lane prior_num, we set 192 lane priors in our experiments. This can change in config.  And for params (x,y,theta), they are used to determine the lane shape and extract lane feature, they are progressively refined when inference in different stages to get more precise lane shape, so they are dynamic in different stages.  Unlike line anchors set in Line-CNN and laneatt, they are some fixed predefined lines in image, so they need huge number of anchors(eg: 1000) in case to cover almost situations.

Q2: We regress (x, y, theta) and corresponding 72 point x-coordinates together in all stages, but in stage-1 and stage-2 the point coordinates regression is not actually used, they are only used to help regress the (x, y, theta) params. In last stage, the output coordinates are served as detection results.

And for lane feature extraction, as we know, one line can be defined by these params (x, y and theta) right. We uniformly sample `np` points on this line and these extracted points' feature serve as line feature. As these params are refined in every stage, so we can extract more precise features during regression.

Q3: We add two branchs after roi-feature, one branch is for classification, another branch is for regression (x, y, theta, 72 coordinates). Each branch consists of two fcs to compute different feature for cls and reg.

Q4: We concat the features along channel dimensions, and we use 9x1 conv to fuse and enhance feature representations. In this way we can collect features from nearby points and previous layer.

Q5: This is set in training samples selection period, if top-k set to 1, then each ground truth lane will be assigned with one lane prior.  You can refer to DETR bipartite matching process.

Details can be found in our code, you can also find answers in code if I do not explain clearly, we will release our code as soon as possible, we are still preparing and organize our code.

**Comment by ChenzrMax** [comment]
Thank you for your answer. Basically consistent with my idea. Hope for your answer about the tensorrt question.

**Comment by LukeAI** [comment]
@ChenzrMax  @MySuperSoul

I am trying to understand the output of the network so that I can write the post-processing in c++. I'd be happy to share my code once completed.

So, the output is a tensor with dimensions 1 x 192 x 78

I understand that each of the 192 rows, corresponds to a lane prediction. From studying the code, I gather that:

column 1 (the second column) is confidence score and that the last 72 columns (col 6 - col 77) are normalized x-coordinates sampled evenly.

Please could you let me know what Columns 0, 2, 3, 4, 5 are?

**Comment by ChenzrMax** [comment]
@LukeAI You can find the answer in paragraph 3.1 in paper. 0, 1--foreground and background probabilities. 2--length of the lane piror. 3,4,5--the start point and the angle of the lane piror(termed as x,y,Θ).

**Comment by gebawe** [comment]
@MySuperSoul @LukeAI @ChenzrMax I think the order of features per lane written on Sec 3.1 in the paper is misleading. It should be consistent with the code. 
### The correct order of the feature index are:
` 0: background, 
  1: foreground, 
   2: start_y, 
   3: start_x, 
   4: theta, 
   5: length, 
   6 - 77: lane x coordinates
`
